library/alloc/src/string.rs

```diff
@@ -2444,6 +2452,49 @@ impl ToString for i8 {
     }
 }
 
+#[cfg(not(no_global_oom_handling))]
+#[stable(feature = "u16_to_string_specialization", since = "0.00.0")]
+impl ToString for u16 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut buf = String::with_capacity(u16::MAX.log10() as usize + 1);
+        fmt_with_buf(&mut buf, self);
+        buf
+    }
+}
+
+#[cfg(not(no_global_oom_handling))]
+#[stable(feature = "i16_to_string_specialization", since = "0.00.0")]
+impl ToString for i16 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut buf = String::with_capacity(
+            i16::MAX.log10() as usize + 1 /* and incorporate the possible minus sign: */ + 1,
+        );
+        fmt_with_buf(&mut buf, self);
+        buf
+    }
+}
+
+macro_rules! impl_NumToString {
+    ($($t:ident),*) => {
+        $(
+            #[cfg(not(no_global_oom_handling))]
+            #[stable(feature = "num_to_string_specialization", since = "0.00.0")]
+            impl ToString for $t {
+                #[inline]
+                fn to_string(&self) -> String {
+                    let mut buf = String::with_capacity(8);
+                    fmt_with_buf(&mut buf, self);
```

Comment:
In which cases do you think reserving 8 bytes will actually help? Doing something "right away" doesn't necessarily make it faster. If a string is empty and you push something, it will reserve those 8 bytes anyway. On the otherhand, if you unconditionally reserve 8 bytes, but then append a 9+ byte long string, or just leave the string empty, doing that 8-byte allocation was a waste.

So by allocating 8 bytes right away, you aren't really optimising the common (which is probably not as common as you think) case (apart from maybe one or two less branches on the first push), but you are heavily pessimizing everything else, because allocations are expensive.

As for to_string, Display::fmt for [integers](https://github.com/rust-lang/rust/blob/473eaa42e9365c47d129f72693b5d163a20cf369/library/core/src/fmt/num.rs#L209) and [floats](https://github.com/rust-lang/rust/blob/473eaa42e9365c47d129f72693b5d163a20cf369/library/core/src/fmt/float.rs#L54) both work by doing the main conversion inside a statically sized buffer, then just passing writing the result into the string. While there are arguably some improvements which could be made for negative numbers with more than 7 digits (due to reallocating), these changes don't fix that anyway.

Reply:
I will just remove the `impl_NumToString`.